soroban-env-host/src/host/diagnostics_helper.rs

```diff
@@ -0,0 +1,111 @@
+use soroban_env_common::{
+    xdr::{ContractEventType, Hash, ScBytes},
+    BytesObject, EnvBase, Symbol, SymbolSmall, VecObject,
+};
+
+use crate::{budget::AsBudget, Host, HostError, RawVal};
+use crate::{
+    events::{InternalContractEvent, InternalEvent},
+    host_object::HostVec,
+};
+
+use super::Frame;
+
+/// None of these functions are metered, which is why they're behind the is_debug check
+impl Host {
+    fn hash_to_bytesobj(&self, hash: &Hash) -> Result<BytesObject, HostError> {
+        self.add_host_object::<ScBytes>(hash.as_slice().to_vec().try_into()?)
+    }
+
+    // Will not return error if frame is missing
+    fn get_current_contract_id(&self) -> Result<Option<Hash>, HostError> {
+        self.with_current_frame_opt(|frame| match frame {
+            #[cfg(feature = "vm")]
+            Some(Frame::ContractVM(vm, _, _)) => Ok(Some(vm.contract_id.clone())),
+            Some(Frame::HostFunction(_)) => Ok(None),
+            Some(Frame::Token(id, _, _)) => Ok(Some(id.clone())),
+            #[cfg(any(test, feature = "testutils"))]
+            Some(Frame::TestContract(tc)) => Ok(Some(tc.id.clone())),
+            None => Ok(None),
+        })
+    }
+
+    fn record_system_debug_contract_event(
+        &self,
+        type_: ContractEventType,
+        contract_id: Option<BytesObject>,
+        topics: VecObject,
+        data: RawVal,
+    ) -> Result<(), HostError> {
+        let ce = InternalContractEvent {
+            type_,
+            contract_id,
+            topics,
+            data,
+        };
+        self.get_events_mut(|events| {
+            Ok(events.record(InternalEvent::StructuredDebug(ce), self.as_budget()))
+        })?
+    }
+
+    // Emits an event with topic = ["fn_call", called_contract_id, function_name] and
+    // data = [arg1, args2, ...]
+    // Should called prior to opening a frame for the next call so the calling contract can be inferred correctly
+    pub fn fn_call_diagnostics(
+        &self,
+        called_contract_id: &Hash,
+        func: &Symbol,
+        args: &[RawVal],
+    ) -> Result<(), HostError> {
+        if !self.is_debug() {
+            return Ok(());
+        }
+
+        let mut calling_contract: Option<BytesObject> = None;
+        if let Some(calling_hash) = self.get_current_contract_id()? {
+            calling_contract = Some(self.hash_to_bytesobj(&calling_hash)?);
+        }
+
+        self.as_budget().with_free_budget(|| {
+            let topics: Vec<RawVal> = vec![
+                SymbolSmall::try_from_str("fn_call")?.into(),
+                self.hash_to_bytesobj(called_contract_id)?.into(),
+                func.into(),
+            ];
+
+            self.record_system_debug_contract_event(
```

Comment:
I see we do a roundtrip of values->hostvec->values here (e.g. for topics). Given that this is internal, can we avoid it, at least where it's easy (like in the vec case). An `.._internal` helper can be introduced in case if there are external users of this function.

Reply:
I don't think I understand what you're proposing here. Are you saying I can avoid the conversion to HostVec?

Reply:
Exactly. In general, in the host code we should try to avoid creating the host objects when possible. I didn't try to find every code path that uses these functions though, so the refactoring might be trickier than I think.

Reply:
I'm not sure I agree -- I don't love adding host objects in all cases, but I think this case is _plausibly_ appropriate / playing to their strengths:

- Host object construction is automatically metered
- Host objects that are containers (maps and vecs) store key and value types that are inherently polymorphic: `RawVal` (so can contain a mixture of things like symbols, numbers, strings, hashes)
- Host objects (once constructed) are subsequently _referenced_ by integer handles stored in RawVals, and those handles/RawVals are inherently both cheap to clone (cheaper even than Rc<> or a weak pointer, much cheaper than a pointed-to actual vector object) and actually constant-time to clone (so cloning a Vector of them is inherently a shallow operation -- doesn't clone the pointed-to Host Objects)

It's also _plausibly_ a waste to use a host object here. Some counterarguments might be:

  - That this is really a fixed-size array, not so much a "vector"
  - And it's small, only 3 elements
  - And we're going to allocate lots of them over and over

And in that case it might even make sense to allocate it in-place, inside the event, using a `[Rawval;3]` or a `TinyVec<[RawVal;3]>` because copying 3*8=24 bytes is certainly cheaper than hitting malloc at all and we might not be duplicating the resulting event object all that often.

But I would not go so far as to say "in general we should avoid host objects". We should use them when they're appropriate, they have unique properties relative to normal rust objects.

Reply:
Sure, I don't say there is something wrong with the host objects in general. However, creating a host object in host without passing it back to guest has bad smell to me. The reason is that since host objects aren't really usable themselves we have to access the underlying Rust object if we want to do something interesting about the object (besides copy). So a lot of the times what we ended up doing was a roundtrip of 'convert rust object to host object (clone)'->'call function that operates on a host object'->'access the underlying rust object', instead of just 'call function that operates on rust object' (frequently no clones/allocations are needed at all).

I've already realized that this case is trickier because we also create events coming from the guest side. Depending on the usage patterns this may or may not be less efficient - I've left a comment about this here: https://github.com/stellar/rs-soroban-env/issues/648. I wouldn't block this PR on this in any case.